lib/hammer_cli_foreman/api/oauth/password_grant.rb

```diff
@@ -0,0 +1,82 @@
+require 'hammer_cli_foreman/openid_connect'
+
+module HammerCLIForeman
+  module Api
+    module Oauth
+      class PasswordGrant < ApipieBindings::Authenticators::TokenAuth
+        attr_accessor :openidc_token_endpoint, :client_id, :user, :password, :token
+
+        def initialize(openidc_token_endpoint, client_id, user, password)
+          @openidc_token_endpoint = openidc_token_endpoint
+          @client_id = client_id
+          @user = user
+          @password = password
+          super set_token(openidc_token_endpoint, client_id, user, password)
+        end
+
+        def authenticate(request, token)
+          if HammerCLI.interactive?
+            set_token_interactively
+          end
+          super
+        end
+
+        def set_token_interactively
+          @token ||= set_token(get_openidc_token_endpoint, get_client_id, get_user, get_password)
+        end
+
+        def set_token(input_openidc_token_endpoint, input_client_id, input_user, input_password)
+          @openidc_token_endpoint = input_openidc_token_endpoint if input_openidc_token_endpoint
+          @user = input_user
+          @password = input_password
+          @client_id = input_client_id if input_client_id
+          if @user && @password && @openidc_token_endpoint && @client_id
+            @token = HammerCLIForeman::OpenidConnect.new(
+              @openidc_token_endpoint, @client_id).get_token(@user, @password)
+          else
+            @token = nil
+          end
+        end
+
+        def error(ex)
+          # if ex.is_a?(RestClient::Unauthorized)
```

Comment:
This method is handler of attempt to login with PasswordGrant. If you provide wrong credentials or client id or token endpoint it endsup here with RestClient::InternalServerError. We need to catch it and and change to UnauthorizedError with message useful to the user in his context.
We were not able to get the token and sent empty server returned "Nil JSON web token" and it is not useful unless you know what is going on under the hood. Similar it should be for code grant.